test/parser.js

```diff
@@ -1,20 +1,112 @@
+const { expect } = require('chai')
 
-var it = require('mocha').it
-var expect = require('chai').expect
-var describe = require('mocha').describe
+const posthtml = require('../lib')
 
-var parser = require('posthtml-parser')
-var render = require('posthtml-render')
+const input = '<div class="button"><rect /><div class="button__text">Text</div></div>'
 
-var html = fs.readFileSync(
-  path.resolve(__dirname, 'templates/parser.html'), 'utf8'
-)
+function test (html, done) {
+  const options = { singleTags: ['rect'], closingSingleTag: 'slash' }
+  posthtml()
+    .process(html, options)
+    .then(result => {
+      expect(input).to.eql(result.html)
+      done()
+    })
+    .catch(error => done(error))
+}
 
-describe('Parser', function () {
-  it('parser => render', function (done) {
-    expect(html).to.eql(render(parser(html)))
-    done()
+describe('Set options', () => {
+  it('html equal', done => {
+    test(input, done)
+  })
+})
+
+describe('Skip html parsing & ', () => {
+  const options = { skipParse: true }
+
+  it('use number tree.', done => {
+    const tree = 123456789
+    expect(posthtml()
+      .process(tree, options)
+      .then(({ html }) => {
+        expect('123456789').to.eql(html)
+        done()
+      })
+      .catch(error => done(error))
+    )
+  })
+
+  it('use string tree.', done => {
+    const tree = '123456789'
+    expect(posthtml()
+      .process(tree, options)
+      .then(({ html }) => {
+        expect('123456789').to.eql(html)
+        done()
+      })
+      .catch(error => done(error))
+    )
+  })
+})
+
+describe('Skip html parsing & use tree from options.', () => {
+  const options = { singleTags: ['rect'], closingSingleTag: 'slash' }
+  const tree = [
+    {
+      tag: 'div',
+      attrs: { class: 'button' },
+      content: [
+        { tag: 'rect' },
+        {
+          attrs: { class: 'button__text' },
+          content: ['Text']
+        }
+      ]
+    }
+  ]
+
+  it('Set use tree', done => {
+    options.skipParse = true
+    expect(posthtml()
+      .process(tree, options)
+      .then(({ html }) => {
+        expect(input).to.eql(html)
+        done()
+      })
+      .catch(error => done(error))
+    )
+  })
+})
+
+describe('Set option', () => {
+  const options = { singleTags: ['rect'], closingSingleTag: 'slash' }
+  const html = '<?php echo "Hello word"; ?>'
+  const multiHTML = `<!doctype><html><body>${html}</body></html>`
+
+  options.directives = [
+    { name: '?php', start: '<', end: '>' }
+  ]
+
+  it('directive ?php', done => {
+    expect(posthtml()
+      .process(html, options)
+      .then(result => {
+        expect(html).to.eql(result.html)
+        done()
+      })
+      .catch(error => done(error))
+    )
+  })
+
+  it('directive ?php with multi html', done => {
+    expect(posthtml()
+      .process(multiHTML, options)
+      .then(result => {
+        expect(multiHTML).to.eql(result.html)
+        done()
+      })
+      .catch(error => done(error))
+    )
```

Comment:
Something went wrong here

Reply:
Yeah, you are right! Something went wrong with my own written script. I will rebase the PR.